DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h

```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
-    if (literals.size()) {
-      literalSize = (literals.size() * sizeof(STYP)) / sizeof(stream_t) + (sizeof(STYP) < sizeof(stream_t));
-      expandStorage(literalSize);
-      bl->storeLiterals(literalSize, reinterpret_cast<const stream_t*>(literals.data()));
-    }
-    *meta = Metadata{messageLength, literals.size(), sizeof(uint64_t), sizeof(stream_t), static_cast<uint8_t>(encoder->getSymbolTablePrecision()), opt,
-                     encoder->getMinSymbol(), encoder->getMaxSymbol(), dictSize, dataSize, literalSize};
+    // store incompressible symbols if any
+    const size_t nLiteralSymbols = [&]() {
+      const size_t nSymbols = literals.size();
+      if (!literals.empty()) {
+        // introduce padding in case literals don't align;
+        const size_t nPaddingElements = (nSymbols * sizeof(input_t)) % sizeof(storageBuffer_t);
+        literals.insert(literals.end(), nPaddingElements, {});
+
+        const size_t literalStorageSize = (literals.size() * sizeof(input_t)) / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
+        expandStorage(literalStorageSize);
+        thisBlock->storeLiterals(literalStorageSize, reinterpret_cast<const storageBuffer_t*>(literals.data()));
+      }
+      return nSymbols;
+    }();
+
+    *thisMetadata = Metadata{messageLength,
+                             literals.size(),
+                             sizeof(ransState_t),
+                             sizeof(ransStream_t),
+                             static_cast<uint8_t>(encoder->getSymbolTablePrecision()),
+                             opt,
+                             encoder->getMinSymbol(),
+                             encoder->getMaxSymbol(),
+                             static_cast<int32_t>(frequencyTable.size()),
+                             dataSize,
+                             static_cast<int32_t>(nLiteralSymbols)};
 
   } else { // store original data w/o EEncoding
-    const size_t szb = messageLength * sizeof(STYP);
-    const int dataSize = szb / sizeof(stream_t) + (sizeof(STYP) < sizeof(stream_t));
+    const size_t messageSizeBytes = messageLength * sizeof(input_t);
+    const int dataSize = messageSizeBytes / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
     // no dictionary needed
     expandStorage(dataSize);
-    *meta = Metadata{messageLength, 0, sizeof(uint64_t), sizeof(stream_t), probabilityBits, opt, 0, 0, 0, dataSize, 0};
-    //FIXME: no we don't need an intermediate vector.
+    const size_t nPaddingElements = messageSizeBytes % sizeof(storageBuffer_t);
+    //FIXME(milettri): we should be able to do without an intermediate vector;
     // provided iterator is not necessarily pointer, need to use intermediate vector!!!
-    std::vector<STYP> vtmp(srcBegin, srcEnd);
-    bl->storeData(meta->nDataWords, reinterpret_cast<const W*>(vtmp.data()));
+    std::vector<input_t> vtmp(srcBegin, srcEnd);
+    vtmp.insert(vtmp.end(), nPaddingElements, {});
```

Comment:
Since vtmp is a local vector with size given in constructor, `insert` will certainly lead to reallocation.
That's why to avoid it I calculate the padded size in the beginning, create a vector of the final size, then 
copy the data, the padding by 0s is there by construction, see https://github.com/AliceO2Group/AliceO2/blob/1e054d67466173870d49c8726f955f90de4f4f97/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h#L861-L869